src/linalg/lmul.jl

```diff
@@ -0,0 +1,233 @@
+                                                <:AbstractStridedLayout, T}) where {UNIT,T<:BlasComplex}
+A,x = M.A,M.B
+BLAS.trmm!('L', 'L', 'C', UNIT, one(T), triangulardata(A)', x)
+end
+
+
+function materialize!(M::MatLmulMat{<:TriangularLayout})
+    A,X = M.A,M.B
+    size(A,2) == size(X,1) || thow(DimensionMismatch("Dimensions must match"))
+    for j in axes(X,2)
+        apply!(*, A, view(X,:,j))
+    end
+    X
+end
+
+@inline function materialize!(M::BlasMatRmulMat{<:AbstractStridedLayout,
+                                                <:TriangularLayout{UPLO,UNIT,<:AbstractColumnMajor},T}) where {UPLO,UNIT,T<:BlasFloat}
+    x,A = M.A,M.B
+    BLAS.trmm!('R', UPLO, 'N', UNIT, one(T), triangulardata(A), x)
+end
+
+@inline function materialize!(M::BlasMatRmulMat{<:AbstractStridedLayout,
+                                                <:TriangularLayout{'L',UNIT,<:AbstractRowMajor},T}) where {UNIT,T<:BlasFloat}
+    x,A = M.A,M.B
+    BLAS.trmm!('R', 'U', 'T', UNIT, one(T), transpose(triangulardata(A)), x)
+end
+
+@inline function materialize!(M::BlasMatRmulMat{<:AbstractStridedLayout,
+                                                <:TriangularLayout{'U',UNIT,<:AbstractRowMajor},T}) where {UNIT,T<:BlasFloat}
+x,A = M.A,M.B
+BLAS.trmm!('R', 'L', 'T', UNIT, one(T), transpose(triangulardata(A)), x)
+end
+
+@inline function materialize!(M::BlasMatRmulMat{<:AbstractStridedLayout,
+                                                <:TriangularLayout{'L',UNIT,<:ConjLayout{<:AbstractRowMajor}},T}) where {UPLO,UNIT,T<:BlasComplex}
+    x,A = M.A,M.B
+    BLAS.trmm!('R', 'U', 'C', UNIT, one(T), triangulardata(A)', x)
+end
+
+@inline function materialize!(M::BlasMatRmulMat{<:AbstractStridedLayout,
+                                                <:TriangularLayout{'U',UNIT,<:ConjLayout{<:AbstractRowMajor}},T}) where {UPLO,UNIT,T<:BlasComplex}
+x,A = M.A,M.B
+BLAS.trmm!('L', 'U', 'C', UNIT, one(T), triangulardata(A)', x)
+end
+
+
+function materialize!(M::MatRmulMat{<:AbstractStridedLayout,<:TriangularLayout})
+    A,X = M.A,M.B
+    apply!(*, X', A')
+    A
+end
+
+
+####
+# Diagonal
+####
+mulapplystyle(::DiagonalLayout, ::DiagonalLayout) = LmulStyle()
+
+mulapplystyle(::DiagonalLayout, _) = LmulStyle()
+mulapplystyle(_, ::DiagonalLayout) = RmulStyle()
+
+# Diagonal multiplication never changes structure
+similar(M::Lmul{<:DiagonalLayout}, ::Type{T}, axes) where T = similar(M.B, T, axes)
+# equivalent to rescaling
+function materialize!(M::Lmul{<:DiagonalLayout{<:AbstractFillLayout}})
+    M.B .= getindex_value(M.A.diag) .* M.B
+    M.B
+end
+
+copy(M::Lmul{<:DiagonalLayout,<:DiagonalLayout}) = Diagonal(M.A.diag .* M.B.diag)
+copy(M::Lmul{<:DiagonalLayout}) = M.A.diag .* M.B
+copy(M::Lmul{<:DiagonalLayout{<:AbstractFillLayout}}) = getindex_value(M.A.diag) .* M.B
+copy(M::Lmul{<:DiagonalLayout{<:AbstractFillLayout},<:DiagonalLayout}) = Diagonal(getindex_value(M.A.diag) .* M.B.diag)
```

Comment:
Why not define `materialize!` and let the default `copy(M::Lmul)` do the job? (Same question for `Rmul`)

Reply:
Because in InfiniteArrays.jl/InfiniteLinearAlgebra.jl `M.A` and `M.B` are ∞-dimensional, and these result in a lazy `BroadcastArray`

Reply:
And the special case for `FillLayout` is because if they are ∞-dimensional banded matrices scalar multiplication lowers down. So we can do:
```julia
A = _BandedMatrix((1:∞)', ∞, -1, 1)
2.0*A === _BandedMatrix((2.0:2.0:∞)', ∞, -1, 1)
```
which is useful for fast building of spectral methods.